File: java-compat/src/Ice/src/main/java/IceInternal/SequencePatcher.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

package IceInternal;

public class SequencePatcher implements Ice.ReadValueCallback
{
    public SequencePatcher(java.lang.Object[] seq, Class<?> cls, int index)
    {
        _seq = seq;
        _cls = cls;
        _index = index;
    }

    public void valueReady(Ice.Object v)
    {
        if(v != null)
        {
            //
            // Raise ClassCastException if the element doesn't match the expected type.
            //
            if(!_cls.isInstance(v))
            {
                throw new ClassCastException("expected element of type " + _cls.getName() + " but received " +
                                             v.getClass().getName());
            }
        }

        _seq[_index] = v;
    }

    private java.lang.Object[] _seq;
    private Class<?> _cls;
    private int _index;
}

File: cs/src/Ice/SequencePatcher.cs
// **********************************************************************
//
// Copyright (c) 2003
// ZeroC, Inc.
// Billerica, MA, USA
//
// All Rights Reserved.
//
// Ice is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License version 2 as published by
// the Free Software Foundation.
//
// **********************************************************************

namespace IceInternal
{

public class SequencePatcher : Patcher
{
    public
    SequencePatcher(System.Object[] seq, System.Type cls, string type, int index)
    {
	_seq = seq;
	_cls = cls;
	_type = type;
	_index = index;
    }
    
    public virtual void
    patch(Ice.Object v)
    {
	//
	// Raise ClassCastException if the element doesn't match the expected type.
	//
	if (!_cls.IsInstanceOfType(v.GetType()))
	{
	    throw new System.InvalidCastException("expected element of type " + _cls.FullName + " but received " + v.GetType().FullName);
	}
	
	_seq[_index] = v;
    }
    
    public virtual string
    type()
    {
	return _type;
    }
    
    private System.Object[] _seq;
    private System.Type _cls;
    private string _type;
    private int _index;
}

}
